# **Funções**

São blocos de código que executam funcionalidades específicas.

Normalmente são utilizados para evitar que determinada parte do seu código seja escrito várias vezes.

Em Python sua sintaxe é definida usando ***def*** e atribuindo um nome a ela, veja um exemplo:

In [ ]:
def funcao():
    print("Bloco de código")


Observando essa função, podemos extrair algumas informações, iniciando com a palavra reservada para funções ***def***, o nome atribuido à função *funcao* e os parênteses **( )** utilizado para definição dos dados de entrada da função, também chamados de parâmetros.

Em seguida usa-se dois pontos : e abaixo o bloco de código a ser executado, que neste caso é apenas imprimir de uma string.

Para “chamar” uma função, utilizamos o nome que foi definido, dessa forma:

In [ ]:
funcao()

Bloco de código


# **Parâmetros**

Além de executar código, funções também podem receber e retornar dados.

Podemos enviar dados para uma função através de seus parâmetros.

Observe o exemplo:

In [ ]:
def imprime_nome(nome):
    print(f"Nome: {nome}")

##f - string -> tudo que estiver entre {} ele irá substituir


imprime_nome("Erickson")
imprime_nome("Renan")
imprime_nome("Daniel")

Nome: Erickson
Nome: Renan
Nome: Daniel


No **print** acima usamos a notação de formatação mais atual do Python, a **F-strings**.

Quando a função é chamada, passamos uma string como dado de entrada (através do parâmetro **nome**), que é concatenada e impressa dentro da função.

Caso nenhum valor seja infomado ao chamar a função, um erro será gerado. Por exemplo, o seguinte código:

In [ ]:
def imprime_nome(nome):
    print(f"Nome: {nome}")


imprime_nome()

TypeError: imprime_nome() missing 1 required positional argument: 'nome'

Ocasionará o seguinte erro:

***TypeError: imprime_nome() missing 1 required positional argument: 'nome'***


Podemos resolver esse erro utilizando os “Valores Padrão” e é exatamente isso que veremos agora!

# **Valores Padrão (ou Valores Default)**

A utilização dos valores padrão serve para dar um valor quando quem chamou a função não passar nenhum valor para os parâmetros definidos.

Fazemos isso dessa forma:

In [ ]:
def flor(flor='Rosa', cor='Vermelha'):
    print(f"A cor da {flor} é {cor}")


flor()
flor("Azul")

A cor da Rosa é Vermelha
A cor da Azul é Vermelha


Ou seja, o erro anterior não ocorreu novamente!

# **Chamada de Função Posicional versus Chamada de Função Nomeada**

**Chamada de Função Posicional** - Quando chamamos uma função, podemos utilizar a posição padrão dos parâmetros da definição da função para fazer a passagem dos valores das variáveis.

Para entender melhor, veja o exemplo a seguir:

In [ ]:
def monta_computador(cpu='', armazenamento=0, memoria=0):
    print(f'A configuração é: \n\t- CPU: {cpu}\n\t- Armazenamento: {armazenamento}Tb\n\t- Memória: {memoria}Gb')

monta_computador('Intel Core i9',4,64)

A configuração é: 
	- CPU: Intel Core i9
	- Armazenamento: 4Tb
	- Memória: 64Gb


O programador que escreveu a chamada da função monta_computador está respeitando a posição dos parâmetros, ou seja:

O valor** "Intel Core i9"** é referente ao primeiro parâmetro **(cpu)**
O valor **4** é referente ao segundo parâmetro *(armazenamento)*
O valor **64** se refere ao terceiro parâmetro **(memoria)**

Essa é uma chamada de função posicional, ou seja: que respeita a ordem dos parâmetros.

Outra forma de fazer essa chamada de função é utilizar os nomes dos parâmetros!

**Chamada de Função Nomeada** -Aqui não é necessário passar todos os parâmetros e nem mesmo respeitar a ordem de declaração desses parâmetros!

Veja o mesmo exemplo, mas agora utilizando os nomes dos parâmetros:

In [ ]:
monta_computador(memoria=64, armazenamento=4, cpu='Intel Core i9') # Parâmetros fora de ordem.
monta_computador(memoria=1, cpu='Pentium III') # Faltando o parâmetro armazenamento, que assumirá o valor padrão = 0

A configuração é: 
	- CPU: Intel Core i9
	- Armazenamento: 4Tb
	- Memória: 64Gb
A configuração é: 
	- CPU: Pentium III
	- Armazenamento: 0Tb
	- Memória: 1Gb


# **Parâmetros opcionais (*args)**

Caso você queira desenvolver uma função que recebe um número variável de parâmetros, você pode utilizar o asterisco antes do nome da variável, por exemplo:
## *args, *idade , *nome
Observação: O nome ***args*** é uma convenção, ou seja uma boa prática entre programadores Python, contudo, nada te impede de alterar esse nome para *numeros por exemplo.
Dessa forma, a definição da função seria:
def maior_30(*numeros):

### Dessa forma, a função receberá os argumentos em forma de Tupla e você poderá processá-los com um **loop for** por exemplo!

Veja o código abaixo para entender melhor:

In [ ]:
def maior_30(*args): # A variável args é uma tupla que pode ter tamanho 0 (tupla vazia)
    print(args)
    print(type(args))

    for num in args:
        if num > 30:
            print(num)

print("Chamando a função com 6 argumentos")
maior_30(10, 20, 30, 40, 50, 60)


print("\nChamando a função sem nenhum argumento")
maior_30()

Chamando a função com 6 argumentos
(10, 20, 30, 40, 50, 60)
<class 'tuple'>
40
50
60

Chamando a função sem nenhum argumento
()
<class 'tuple'>


### Exemplo com um argumento obrigatório e argumentos adicionais opcionais

In [ ]:
# Agora exigirá pelo menos 1 objeto, o arg1
def printVarInfo(arg1, *vartuple):

   # Imprimindo o valor do primeiro argumento
    print ("O parâmetro arg1 passado foi: ", arg1)

   # Imprimindo o valor do segundo argumento
    for item in vartuple:
        print ("O parâmetro vartuple passado foi: ", item)
    return;

In [ ]:
# Fazendo chamada à função usando apenas 1 argumento
printVarInfo(10)

O parâmetro arg1 passado foi:  10


In [ ]:
printVarInfo('Chocolate', 'Morango', 'Leite condensado')

O parâmetro arg1 passado foi:  Chocolate
O parâmetro vartuple passado foi:  Morango
O parâmetro vartuple passado foi:  Leite condensado


In [ ]:
# Fazendo chamada à função sem argumento gerará erro
printVarInfo()

TypeError: printVarInfo() missing 1 required positional argument: 'arg1'

# **Parâmetros opcionais nomeados** ***kwargs*

Agora, se quiser desenvolver uma função com número variável de **parâmetros nomeados**, utilize **kwargs.

Dessa forma, todos os dados passados à função serão guardados nessa variável ***kwargs***, em formato de um **dicionário**.

Oberve como podemos obter a chave a valor deles percorrendo os itens deste dicionário:

In [8]:
def dados_pessoa(**kwargs):
    print(type(kwargs))
    print("Tamanho do dicionário =", len(kwargs))

    for chave, valor in kwargs.items():
        print(f"{chave}: {valor}")


print("Chamando a função com 3 argumentos nomeados")
dados_pessoa(nome='João', idade=35, carreira='Desenvolvedor Fullstack')

print("\nChamando a função sem nenhum argumento")
dados_pessoa() # Observe que irá gerar um dict vazio

Chamando a função com 3 argumentos nomeados
<class 'dict'>
Tamanho do dicionário = 3
nome: João
idade: 35
carreira: Desenvolvedor Fullstack

Chamando a função sem nenhum argumento
<class 'dict'>
Tamanho do dicionário = 0


In [9]:
print("Chamando a função com 5 argumentos nomeados")
dados_pessoa(nome='João', idade=35, carreira='Desenvolvedor Fullstack', especialidade='frontend', portifolio='www.github.com/fulano')

Chamando a função com 5 argumentos nomeados
<class 'dict'>
Tamanho do dicionário = 5
nome: João
idade: 35
carreira: Desenvolvedor Fullstack
especialidade: frontend
portifolio: www.github.com/fulano


Observação: O nome ***kwargs*** é uma convenção, ou seja uma boa prática entre programadores Python! Contudo, nada te impede de alterar esse nome para **pessoa** por exemplo.

# **Funções com retorno de dados**

As funções também podem retornar valores através da palavra reservada *return*.

Veja o exemplo:

In [ ]:
def soma_dois_numeros(valor1, valor2):
    soma = valor1 + valor2
    return soma

# A variável valor_soma receberá o resultado passado por return (soma)
valor_soma = soma_dois_numeros(32, 15)
print(valor_soma)

# Podemos chamar a função dentro do print diretamente
print(soma_dois_numeros(50, 10))

47
60


# **Funções com retorno múltiplos**

Funções também podem retornar múltiplos dados. Veja o exemplo:

In [ ]:
def soma_dois_numeros_e_calcula_media(valor1, valor2):
    soma = valor1 + valor2
    media = (valor1 + valor2)/2

    return soma, media


valor_soma = soma_dois_numeros_e_calcula_media(32, 15)
print(valor_soma)

print(soma_dois_numeros_e_calcula_media(50, 10))

(47, 23.5)
(60, 30.0)


### Observe que a função acima retornou uma tupla com os dois números.

Podemos associar os reusltados diretamente à duas variáveis, se acharmos conveniente. Veja no exemplo abaixo.

In [ ]:
result_soma, result_media = soma_dois_numeros_e_calcula_media(32, 15)
print("A soma é",result_soma,"e a média é", result_media)

A soma é 47 e a média é 23.5


# **Palavra reservada pass**

Caso você deseje definir uma função sem corpo nenhum, ou seja, sem código, saiba que isso irá disparar o erro ***IndentationError***, pois funções não podem estar vazias.

Porém, se precisar declarar uma função sem nenhum código, use a palavra reservada **pass**, da seguinte forma:

In [ ]:
def funcao1():
    pass

# **Função de uma linha**

Python possibilita a criação de funções com apenas uma linha de código. Veja os exemplo a seguir:

In [ ]:
# Definição das funções
def soma(valor1, valor2): return valor1 + valor2
def divisao(valor1, valor2): return valor1 / valor2
def multiplicacao(valor1, valor2): return valor1 * valor2

# Chamada das funções
print(soma(1, 5))
print(divisao(8, 2))
print(multiplicacao(8, 2))

# **Função lambda**

As funções ***lambda*** são funções anônimas que permitem criar uma função em uma única expressão. Elas são úteis quando você precisa de uma função temporária para uma operação específica, como uma operação matemática ou uma filtragem de listas. As funções lambda são de ordem superior, o que significa que elas podem ser passadas como argumentos para outras funções, como "map()" e "filter()"

In [ ]:
# Definindo uma expressão lambda (função anônima)
# lambda cria uma função em tempo de execução
potencia = lambda num: num ** 2

In [ ]:
potencia(5)

25

In [ ]:
type(potencia)

function

### Para exemplificar como usar a função *lambda*, iremos refazer uma função de uma linha abaixo, porém usando a função *lambda*.
Faça também os outros dois exemplos já feitos para treinar.

In [ ]:
#def soma(valor1, valor2): return valor1 + valor2

soma = lambda valor1, valor2: valor1 + valor2

print(soma(10,20))

30


In [11]:
#def divisao(valor1, valor2): return valor1 / valor2

divisao = lambda valor1, valor2: valor1/valor2

print(divisao(16,8))

2.0


# **Escopo de Variável  - Local e Global**

In [ ]:
# Variável Global
var_global = 10  # Esta é uma variável global

# Função
def multiplica_numeros(num1, num2):
    var_global = num1 * num2  # Esta é uma variável local
    print(var_global)

In [ ]:
multiplica_numeros(5, 25)

125


Observe que var_global que está dentro da função não teve interferência da variável global definida anteriomente, retornou 125 como esperado.
A variável var_global **fora da função** entretanto, continua valendo 10. Veja:

In [ ]:
print(var_global)

10


Não é uma boa prática de programação utilizar o mesmo nome como foi feito acima.
O melhor seria rescrever a função como abaixo.

In [ ]:
# Variável Global
var_global = 10  # Esta é uma variável global

# Função
def multiplica_numeros(num1, num2):
    var_local = num1 * num2   # Esta é uma variável local
    print(var_local)

In [ ]:
multiplica_numeros(5, 25)

125


In [ ]:
# A variável local só é definida dentro da função
# Uma chamada à essa variável fora da função geraria um erro, faça o teste abaixo.

# print(var_local)

É possível usar funções dentro da definição da sua função.
### Criando Funções Usando Outras Funções

In [ ]:
import math

# Verificando se um número é primo
def numPrimo(num):
    if (num % 2) == 0 and num > 2:
        return "Este número não é primo"
    for i in range(3, int(math.sqrt(num)) + 1, 2):
        if (num % i) == 0:
            return "Este número não é primo"
    return "Este número é primo"

In [ ]:
numPrimo(541)

'Este número é primo'

In [ ]:
numPrimo(20)

'Este número não é primo'